File: modeler/cadf.r
library(RMySQL)
library(xts)
library("CADFtest")

################################################################################
# Cointegration Pairs
# I / RB
# M / RM
# Y / P
rm(list=ls())
connect <- dbConnect(MySQL(), host="127.0.0.1", port=3306, dbname = "aq", username="root", password="!QAZ2wsx#EDC")
d <- dbGetQuery(connect, paste0("SELECT d1.datetime as dt, d1.close as d1_c, d2.close as d2_c FROM
                                 (SELECT code, datetime, close FROM future_trade WHERE code='I' AND type='1') d1, 
                                 (SELECT code, datetime, close FROM future_trade WHERE code='RB' AND type='1') d2 
                                 WHERE d1.datetime=d2.datetime"))
dbDisconnect(connect)
################################################################################
d$dt <- as.POSIXlt(d$dt)
d$d1_rtn <- c(0, diff(d$d1_c))/d$d1_c
d$d2_rtn <- c(0, diff(d$d2_c))/d$d2_c
#d <- xts(d[,-1], order.by=d[,1])
plot(d$d1_rtn, d$d2_rtn)
cadftest <- CADFtest(d$d1_rtn~d$d2_rtn, d)
print(cadftest)
################################################################################

File: modeler/cadf.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import pandas as pd
from mysql import connector
import numpy as np
import pandas as pd
import datetime
import matplotlib.pyplot as plt
import matplotlib.dates as mdates

import pprint
import statsmodels.tsa.stattools as ts

class CADF:

    def __init__(self, aq):
        self.aq = aq

    def execute(self):
        self.aq.log("Start")

        code = "P"
        type = "d"
        mysql_connector = connector.connect(host=self.aq.DB_HOST, database=self.aq.DB_NAME,
                                            user=self.aq.DB_USR, password=self.aq.DB_PWD)
        data = pd.read_sql("""SELECT datetime, open, high, low, close, volume, oi FROM future_trade
                            WHERE code='%s' AND type='%s' """ % (code, type), con=mysql_connector)
        mysql_connector.close()

        data["rtn"] = (data.close.diff()/data.close.shift(1))

        self.aq.log("Stop")